### PR TITLE
Add ldap backend support into sudo

### PR DESCRIPTION
**Pull request by AZaugg**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [X] The toolchain has been rebuilt successfully (or no changes were made to it)
- [X] The toolchain/worker package manifests are up-to-date
- [X] Any updated packages successfully build (or no packages were changed)
- [X] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [X] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [X] All package sources are available
- [X] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [X] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [X] All source files have up-to-date hashes in the `*.signatures.json` files
- [X] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [X] Documentation has been updated to match any changes to the build system
- [X] Ready to merge

---

###### Summary 

Add the config option to sudo build to allow configuration of sudo via LDAP.

###### Change Log  

- Add openldap to buildrequires and adding --with-ldap option to sudo SPEC file

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Updated SPEC file with config option
- Ran a rebuild of the package 
```sudo  make build-packages -j$(nproc) REBUILD_TOOLS=y SRPM_PACK_LIST="sudo" PACKAGE_REBUILD_LIST="sudo" REFRESH_WORKER_CHROOT=n
```

### CODE REVIEW

File: SPECS/sudo/sudo.spec [hunk 1]
```diff
@@ -36,6 +37,8 @@ the ability to run some (or all) commands as root or another user while logging
     --docdir=%{_docdir}/%{name}-%{version} \
     --with-env-editor \
     --with-pam \
+    --with-ldap \
+    --with-ldap-conf-file=/etc/sudo-ldap.conf \
```
Comment (ddstreet) [inline SPECS/sudo/sudo.spec:41]: is there a reason/need for defaulting to a separate ldap conf file here? Or just copying how Debian/Ubuntu does it?

Comment (AZaugg) [reply SPECS/sudo/sudo.spec:41]: I essentially followed what was on the man page [here](https://linux.die.net/man/5/sudoers.ldap).

**Comment by AZaugg** [comment]
> Can you provide some detail about how you actually tested this to make sure it works and does what you expect?

I added the options to the sudo.spec file, and built out the RPM, i then upgraded the local RPM and ensured the the config options were set:

```
azaugg@azaugg-ld77 [ ~ ]$ rpm -qa|grep sudo
sudo-1.9.13p3-1.cm2.x86_64
azaugg@azaugg-ld77 [ ~ ]$ sudo sudo -V
Sudo version 1.9.13p3
Configure options: --prefix=/usr --bindir=/usr/bin --libexecdir=/usr/lib --docdir=/usr/share/doc/sudo-1.9.13p3 --with-env-editor --with-pam --with-linux-audit --enable-zlib=system --with-passprompt=[sudo] password for %p:
```

Now after upgrade

```
azaugg@azaugg-ld77 [ ~ ]$ sudo rpm -U sudo-1.9.13p3-2.cm2.x86_64.rpm
/sbin/ldconfig: /lib/libwbclient.so.0 is not a symbolic link

/sbin/ldconfig: /lib/libwbclient.so.0 is not a symbolic link

azaugg@azaugg-ld77 [ ~ ]$ rpm -qa|grep sudo
sudo-1.9.13p3-2.cm2.x86_64
azaugg@azaugg-ld77 [ ~ ]$ sudo sudo -V
Sudo version 1.9.13p3
Configure options: --prefix=/usr --bindir=/usr/bin --libexecdir=/usr/lib --docdir=/usr/share/doc/sudo-1.9.13p3 --with-env-editor --with-pam --with-ldap --with-linux-audit --enable-zlib=system --with-passprompt=[sudo] password for %p:
```

My other colleague took the RPM and configured the ldap server and was able to bind against the directory server and control sudo via it

File: SPECS/sudo/sudo.spec (hunk 1 above)
Comment (trungams) [reply SPECS/sudo/sudo.spec:41]: The man page also mentions this
> Also note that on systems using the OpenLDAP libraries, default values specified in /etc/openldap/ldap.conf or the user's .ldaprc files are not used.
so I think leaving it to the current path is fine.

**Review by trungams — APPROVED** [review]
Looks good to me!

File: SPECS/sudo/sudo.spec [hunk 2]
```diff
@@ -15,6 +15,7 @@ BuildRequires:  openssl-libs
 BuildRequires:  pam-devel
 BuildRequires:  sed
 BuildRequires:  zlib-devel
+BuildRequires:  openldap
```
Comment (trungams) [inline SPECS/sudo/sudo.spec:18]: The documentation suggests `openldap-devel`: https://www.sudo.ws/docs/install/#linux

File: SPECS/sudo/sudo.spec (hunk 1 above)
Comment (trungams) [reply SPECS/sudo/sudo.spec:41]: There seems to be a slight difference between the man page you linked and the official documentation: https://www.sudo.ws/docs/man/1.9.13/sudoers.ldap.man/#Configuring_ldap.conf. The default value is `/etc/ldap.conf` and it's shared among other ldap clients. Unless users have specific requirements for using sudo with ldap, they don't need to create a separate config file. Let's remove that parameter and leave it to default to `/etc/ldap.conf`

**Review by trungams — CHANGES_REQUESTED** [review]

**Comment by trungams** [comment]
@AZaugg I added some comments. In addition, could you also provide some information on how testing was done to verify sudo works fine with ldap? An example `/etc/ldap.conf` and a brief summary of the steps taken to get sudo working with openldap would be good.

File: SPECS/sudo/sudo.spec (hunk 2 above)
Comment (AZaugg) [reply SPECS/sudo/sudo.spec:18]: Looking at sudo build options when enabling ldap, it requires access to iber.h and ldap.h, I took a look at the mariner-supplied packages, both these headerfiles are provided in openldap

```
azaugg@azaugg-ld77 [ ~/CBL-Mariner/SPECS/sudo ]$ sudo repoquery -l openldap|grep inc
Last metadata expiration check: 0:01:52 ago on Thu May 18 15:26:41 2023.
- Fix CVE-2022-29155
/usr/include/lber.h
/usr/include/lber_types.h
/usr/include/ldap.h
/usr/include/ldap_cdefs.h
/usr/include/ldap_features.h
/usr/include/ldap_schema.h
/usr/include/ldap_utf8.h
/usr/include/ldif.h
/usr/include/openldap.h
/usr/include/slapi-plugin.h
```

Comment (AZaugg) [reply SPECS/sudo/sudo.spec:18]: Hmm, come to think of it, I don't think Mariner ships with a devel package for openldap
```
azaugg@azaugg-ld77 [ ~/CBL-Mariner/SPECS/sudo ]$ sudo yum info openldap-devel
[sudo] password for azaugg:
Loaded plugin: tdnfrepogpgcheck
[using capability match for 'openldap-devel']
Name          : openldap
Arch          : x86_64
Epoch         : 0
Version       : 2.4.57
Release       : 8.cm2
Install Size  :   1.65M 1732185 (1732185)
Repo          : @System
...
...
...
```

File: SPECS/sudo/sudo.spec (hunk 1 above)
Comment (AZaugg) [reply SPECS/sudo/sudo.spec:41]: Sounds good.

File: SPECS/sudo/sudo.spec (hunk 2 above)
Comment (trungams) [reply SPECS/sudo/sudo.spec:18]: Hmm you are right. Our [openldap spec file](https://github.com/microsoft/CBL-Mariner/blob/main/SPECS/openldap/openldap.spec) doesn't specify a devel package. Let me check back on this

Comment (oliviacrain) [reply SPECS/sudo/sudo.spec:18]: I would still use the `openldap-devel` name that's currently attached to the main package. It makes less work for us when we inevitably split the package properly in a future release.

Comment (AZaugg) [reply SPECS/sudo/sudo.spec]: Done, referenced to openldap-devel. I do want to point out that if a devel package is needed, we must have the includes iber.h and ldap.h in the devel file

**Review by trungams — APPROVED** [review]

**Comment by ddstreet** [comment]
@trungams sounds like your investigation into openldap conf shows the separate conf file is not needed (by default), and from @AZaugg comment it sounds like the actual testing with the modified rpm that defaults to the normal ldap conf file does work (if i'm reading the comment right), so it all LGTM, thanks!

**Review by ddstreet — APPROVED** [review]
as mentioned in last comment, LGTM

**Pull request merged**